src/ios/CodePushReportingManager.m

```diff
@@ -1,43 +1,42 @@
 #import "CodePushReportingManager.h"
+#import "StatusReport.h"
 
 @implementation CodePushReportingManager
 
-const NSString* LastVersionPreference = @"CODE_PUSH_LAST_VERSION";
-const NSString* LastVersionDeploymentKeyKey = @"LAST_VERSION_DEPLOYMENT_KEY_KEY";
-const NSString* LastVersionLabelOrAppVersionKey = @"LAST_VERSION_LABEL_OR_APP_VERSION_KEY";
+int HasFailedReport = -1; // -1 = unset, 0 = false, 1 = true
 
-+ (void)reportStatus:(ReportingStatus)status withLabel:(NSString*)label version:(NSString*)version deploymentKey:(NSString*)deploymentKey webView:(UIView*)webView {
+NSString* const FailedStatusReportKey = @"CODE_PUSH_FAILED_STATUS_REPORT_KEY";
+NSString* const LastVersionPreference = @"CODE_PUSH_LAST_VERSION";
+NSString* const LastVersionPreferenceDeploymentKeyKey = @"LAST_VERSION_DEPLOYMENT_KEY_KEY";
+NSString* const LastVersionPreferenceLabelOrAppVersionKey = @"LAST_VERSION_LABEL_OR_APP_VERSION_KEY";
+
++ (void)reportStatus:(StatusReport*)statusReport withWebView:(UIView*)webView {
     @synchronized(self) {
-        if (!deploymentKey) {
+        if (!statusReport.deploymentKey) {
             return;
         }
 
-        NSString* labelParameter = [CodePushReportingManager convertStringParameter:label];
-        NSString* appVersionParameter = [CodePushReportingManager convertStringParameter:version];
-        NSString* deploymentKeyParameter = [CodePushReportingManager convertStringParameter:deploymentKey];
-        NSString* lastVersionLabelOrAppVersionParameter = nil;
-        NSString* lastVersionDeploymentKeyParameter = nil;
-
-        NSUserDefaults* preferences = [NSUserDefaults standardUserDefaults];
-        NSDictionary* lastVersion = [preferences objectForKey:LastVersionPreference];
-        if (lastVersion) {
-            lastVersionLabelOrAppVersionParameter = [CodePushReportingManager convertStringParameter:[lastVersion objectForKey:LastVersionLabelOrAppVersionKey]];
-            lastVersionDeploymentKeyParameter = [CodePushReportingManager convertStringParameter:[lastVersion objectForKey:LastVersionDeploymentKeyKey]];
+        NSString* labelParameter = [CodePushReportingManager convertStringParameter:statusReport.label];
+        NSString* appVersionParameter = [CodePushReportingManager convertStringParameter:statusReport.appVersion];
+        NSString* deploymentKeyParameter = [CodePushReportingManager convertStringParameter:statusReport.deploymentKey];
+        NSString* lastVersionLabelOrAppVersionParameter = statusReport.lastVersionLabelOrAppVersion;
+        NSString* lastVersionDeploymentKeyParameter = statusReport.lastVersionDeploymentKey;
+        if (!lastVersionLabelOrAppVersionParameter || !lastVersionDeploymentKeyParameter) {
```

Comment:
this is doing even more shady "cover-up" kind of stuff. although u can assume that, given your constructor, if one of these is null the other is also probably null as well, but if there's a bug and only one is null now it's harder to tell where the issue is because you replace both anyway